### PR TITLE
Ansible role/playbook for installation of compatibility layer

### PR DESCRIPTION
**Pull request by bedroge**

### CODE REVIEW

File: playbooks/install.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Playbook for installing the EESSI compatibility layer.
+---
+- name: Install compatibility layer
+  hosts: cvmfsstratum0servers
+  environment:
+    PATH: "{{ prefix_location }}/usr/bin:{{ ansible_env.PATH }}"
```
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/install.yml:6]: Definitely makes sense to add /bin, I think I even ran into an issue with bzip2 recently because /bin was not in my path.
I'm not sure if EPREFIX and BPREFIX are really necessary (haven't run into issues), but I think it's indeed better to have them defined anyway. So I'll add/update these variables.

File: playbooks/roles/compatibility-layer/defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+custom_overlay_name: eessi
+custom_overlay_source: git
+custom_overlay_url: https://github.com/EESSI/gentoo-overlay.git
+
+cvmfs_start_transaction: False
+cvmfs_publish_transaction: False
+cvmfs_abort_transaction_on_failures: False
+cvmfs_repository: pilot.eessi-hpc.org
+
+prefix_location: /cvmfs/pilot.eessi-hpc.org/compat/x86_64
```
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/defaults/main.yml:12]: Changed it to `gentoo_prefix_path`.

File: playbooks/roles/compatibility-layer/tasks/add_overlay.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    package: eselect-repository
+    state: present
+
+# We need git in order to add Gentoo overlays hosted on git repositories.
+- name: Install git
+  portage:
+    package: dev-vcs/git
+    state: present
+
+- name: Add our overlay configuration
```
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/add_overlay.yml:25]: There is some protection in the task itself:
```
- name: Add our overlay configuration
  command:
    cmd: "eselect repository add {{ custom_overlay_name }} {{ custom_overlay_source }} {{ custom_overlay_url }}"
    creates: "{{ gentoo_prefix_path }}/etc/portage/repos.conf/eselect-repo.conf"
```
The last line will ensure that this only gets run when there is no eselect-repo.conf yet. This file should have been created if you have run this playbook before. However, if you for instance run this on an Prefix installation where you already added another repo, this task won't run either. But to completely solve this, we would need to check the contents of the file and see if our repo configuration is already in there, which is a bit more tricky. It will have to check if the following section is included in the file, and all settings correspond to the parameters of the playbook:
```
[eessi]
location = /cvmfs/pilot.eessi-hpc.org/test/gentoo/2020/var/db/repos/eessi
sync-type = git
sync-uri = https://github.com/EESSI/gentoo-overlay.git

```

File: playbooks/roles/compatibility-layer/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    path: "{{ prefix_location }}/usr/bin/emerge"
+  register: emerge
+
+- name: Start transaction
+  command: "cvmfs_server transaction {{ cvmfs_repository }}"
+  when: cvmfs_start_transaction
+
+- block:
+  - include_tasks: add_overlay.yml
+    when: emerge.stat.exists
```
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/main.yml:16]: I agree that it would be really nice to repair a broken installation, but I'm not really sure how to do that, since it's hard to find out what broke it in the first place. 

The when clause now only checks if emerge is in your $PATH for the given Prefix path, so basically it only checks if this is a valid Prefix path. If that's not the case, it does not really make sense to continue anyway, so it just skips those included tasks. Maybe it should even abort the playbook in this case? That would allow us to remove the when clauses.

We can go a step further, for instance by doing the check that I described above in a reply to your comment about the `eselect repository add` command, i.e. by checking if our repository is part of `eselect-repo.conf` (and if it's correct). But what if there is another repo in there (which could have been added manually), which could in theory could cause issues? Should we just remove all other repos? I figured that, since this is a very EESSI-specific role anyway and we plan to do everything through this playbook (and only very occasionally), and we only have one custom overlay for now, this check was sufficient for now.

So I'm not really sure how to improve this / automatically fix things; do you have more concrete ideas about this?

**Comment by bedroge** [comment]
Thanks for the review @victorusu. Just made some changes based on your feedback.

File: playbooks/roles/compatibility-layer/tasks/add_overlay.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/add_overlay.yml:25]: By the way, I also tested this by running the playbook multiple times, and for instance only adding a new package to our package set to verify if this gets installed. That worked fine for me. Did you get this error when manually trying to repeat this eselect command, or did it happen while running the playbook?

[the remark this answers is not in the document]
Comment (victorusu) [reply playbooks/roles/compatibility-layer/tasks/add_overlay.yml:25]: I got the error by running manually. 
I haven't noticed the `creates:` section in the task. That's my bad.
This somehow protects the task, but as you said it is not optimal.
What about considering to have a protector task that checks whether the eessi repo is installed.
Something in the lines of

```
- name: Check if eessi repo is installed
   command: eselect repository list -i eessi
   register: eessi_repository_installed

- name: Add eessi overlay configuration
  command:
    cmd: "eselect repository add {{ custom_overlay_name }} {{ custom_overlay_source }} {{ custom_overlay_url }}"
    when: not eessi_repository_installed.stdout.find('eessi')
```

File: playbooks/roles/compatibility-layer/tasks/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (victorusu) [reply playbooks/roles/compatibility-layer/tasks/main.yml:16]: What I meant by broken is related to the imposed configuration or conditions on the ansible-playbook itself. So, there won't be some many things to fix, because the playbook doesn't know much about the gentoo-prefix internals. But that's completely ok.
I was more referring to tasks like the following one
```
- name: Make sets directory
   file:
     path: "{{ gentoo_prefix_path }}/etc/portage/sets"
     state: directory
     mode: '0755'
```
This task would be completely skipped if `emerge` was found in the PATH, but could potentially be problematic if someone by hand modifies the folder permissions. 

There is also another aspect. If someone updates the list of packages to install on the following task, after the prefix was installed, then there would be no point of running the Ansible playbook because the task to install the packages would be completely skipped. While the `portage` module is able to detect if a package was installed or not and would mark it as `ok` if it was already there.
```
  - include_tasks: install_packages.yml
     when: emerge.stat.exists
```

That's more or less what I was talking about the idempotency. So, we could run the ansible-playbook regularly to install new packages and guarantee that it is "sane" (within the ansible-playbook capabilities) at all times.

File: playbooks/roles/compatibility-layer/tasks/add_overlay.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/add_overlay.yml:25]: I think this is indeed a better approach for this check. I'll modify this, thanks!

File: playbooks/roles/compatibility-layer/tasks/main.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/main.yml:16]: Hmm, I don't really understand this, but perhaps we're saying the same thing now ;-)
The playbook will always run these tasks again (and hence, fix the permissions on that sets directory), unless it cannot find emerge at the given Prefix path. This is just a check to verify that the given path is an actual Prefix installation (which always should have emerge?). Without emerge, there's no point in running these tasks, I believe, since almost all of them require emerge or somehow depend on emerge. 
This is what you mean in the end of your comment, right?

[the remark this answers is not in the document]
Comment (victorusu) [reply playbooks/roles/compatibility-layer/tasks/main.yml:16]: Not really. I think that if `emerge` is not found, then the tasks should fail and not be skipped.
Because that highlights a problem.
At the same time, changing the folder permissions for the `"{{ gentoo_prefix_path }}/etc/portage/sets"` folder is independent of the `emerge` command.

[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/main.yml:16]: Yes, failing makes sense (I think I proposed that before as well). I'll modify this and do the emerge check and fail it cannot be found. 

In that case it will still not run a few of the tasks if emerge cannot be found, but I don't think that makes sense anyway. If emerge is not available, I don't really see the point of, for instance, making/modifying that sets directory. This particular task is part of the process of adding an overlay and installing packages from that overlay, and I don't see the point in making that directory if you already know you cannot add the overlay and install packages from it.

[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/main.yml]: Done:

```
- name: Check if a Prefix installation is found at the specified location
  stat:
    path: "{{ gentoo_prefix_path }}/usr/bin/emerge"
  register: emerge

- name: Fail if the given Prefix path does not have emerge
  fail:
    msg: |
         The given Gentoo Prefix path does not seem to be valid,
         because the emerge command cannot be found in {{ gentoo_prefix_path }}/usr/bin.
         Please check it and fix (the path to) the Gentoo Prefix installation.
  when: not emerge.stat.exists
```

For an invalid path/installation, you now get this:
```
TASK [compatibility-layer : Fail if the given Prefix path does not have emerge] ******************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "The given Gentoo Prefix path does not seem to be valid,\nbecause the emerge command cannot be found in /tmp/usr/bin.\nPlease check it and fix (the path to) the Gentoo Prefix installation.\n"}
```

[the remark this answers is not in the document]
Comment (victorusu) [reply playbooks/roles/compatibility-layer/tasks/main.yml]: > This particular task is part of the process of adding an overlay and installing packages from that overlay, and I don't see the point in making that directory if you already know you cannot add the overlay and install packages from it.

Ah! I reviewed it with a different mindset! Sorry about that.
In this case, yeap! You are right. I agree with you too. It is pointless to proceed with other tasks if the `emerge` command is not there.

[the remark this answers is not in the document]
Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/main.yml]: Okay, no problem! Thanks for the feedback.

**Review by victorusu — APPROVED** [review]
`lgtm`

File: playbooks/roles/compatibility-layer/tasks/add_overlay.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  changed_when: False
+
+- name: Add custom overlay configuration
+  command:
+    cmd: "eselect repository add {{ custom_overlay_name }} {{ custom_overlay_source }} {{ custom_overlay_url }}"
+  when: custom_overlay_name not in repositories_installed.stdout
+
+- name: Sync the repositories
+  portage:
+    sync: yes
```
Comment (victorusu) [inline playbooks/roles/compatibility-layer/tasks/add_overlay.yml:37]: One last comment... 😬 
I think this `yes` should be written as `'yes'`, because the `sync` accepts a string, not a boolean (https://docs.ansible.com/ansible/latest/modules/portage_module.html).

Comment (bedroge) [reply playbooks/roles/compatibility-layer/tasks/add_overlay.yml:37]: Nice catch :-)

I would have expected this to work, since the module defines (and hence interprets?) these parameters as strings. Even one of the examples in their documentation uses `sync: yes` without quotes. But I just changed the value to `no`, and then Ansible warns me about converting `False` to `'False'`. So I guess Ansible does first interpret the values (as booleans), and then passes them to the portage modules. 

TLDR: I'll add quotes. Thanks!

File: playbooks/roles/compatibility-layer/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  when: cvmfs_start_transaction
+
+- block:
+  - include_tasks: add_overlay.yml
+
+  - include_tasks: install_packages.yml
+
+  - name: Publish transaction
+    command: "cvmfs_server publish {{ cvmfs_repository }}"
+    when: cvmfs_publish_transaction
```
Comment (cfenoy) [inline playbooks/roles/compatibility-layer/tasks/main.yml:29]:
```suggestion
    when: cvmfs_start_transaction and cvmfs_publish_transaction
```
For completion check that a transaction has been started

File: playbooks/roles/compatibility-layer/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+  rescue:
+    - name: Abort transaction
+      command: "cvmfs_server abort {{ cvmfs_repository }}"
+      when: cvmfs_abort_transaction_on_failures
```
Comment (cfenoy) [inline playbooks/roles/compatibility-layer/tasks/main.yml:34]:
```suggestion
      when: cvmfs_start_transaction and cvmfs_abort_transaction_on_failures
```
For completion check that a transaction has been started

**Review by cfenoy — COMMENTED** [review]
Just for completion the publish transaction and abort transaction tasks should check that a transaction has been started.

**Review by peterstol — APPROVED** [review]
The ansible scripts look good!

Tested on a local prefix with ` ansible-playbook -c local -i localhost, install.yml` after setting cvmfs stuff to false and disabling Become =Yes, which is not required if you run as the same user as the prefix installation.

All TASKS Passed. A minor warning on gathering facts may need fixing later.

The result was a gentoo prefix installation with the EESSI overlay and lmod and archspec packages installed with the 2020.08 set.
 
```
PLAY [Install compatibility layer] **************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************
[WARNING]: error loading fact - please check content
ok: [localhost]

TASK [compatibility-layer : Check if a Prefix installation is found at the specified location] **************************************************
ok: [localhost]

TASK [compatibility-layer : Fail if the given Prefix path does not have emerge] *****************************************************************
skipping: [localhost]

TASK [compatibility-layer : Start transaction] **************************************************************************************************
skipping: [localhost]

TASK [compatibility-layer : include_tasks] ******************************************************************************************************
included: /scistor/gt/psl500/EESSI/compatibility-layer/playbooks/roles/compatibility-layer/tasks/add_overlay.yml for localhost

TASK [compatibility-layer : Install equery command (dependency for the portage module)] *********************************************************
changed: [localhost]

TASK [compatibility-layer : Create repos directory] *********************************************************************************************
changed: [localhost]

TASK [compatibility-layer : Install eselect-repository] *****************************************************************************************
changed: [localhost]

TASK [compatibility-layer : Install git] ********************************************************************************************************
changed: [localhost]

TASK [compatibility-layer : Check which repositories have been installed] ***********************************************************************
ok: [localhost]

TASK [compatibility-layer : Add custom overlay configuration] ***********************************************************************************
changed: [localhost]

TASK [compatibility-layer : Sync the repositories] **********************************************************************************************
ok: [localhost]

TASK [compatibility-layer : Make sets directory] ************************************************************************************************
changed: [localhost]

TASK [compatibility-layer : Make symlink to all package set files in the custom overlay] ********************************************************
changed: [localhost] => (item=2020.08)

TASK [compatibility-layer : include_tasks] ******************************************************************************************************
included: /scistor/gt/psl500/EESSI/compatibility-layer/playbooks/roles/compatibility-layer/tasks/install_packages.yml for localhost

TASK [compatibility-layer : Install package set] ************************************************************************************************
changed: [localhost] => (item=2020.08)

TASK [compatibility-layer : Install additional packages] ****************************************************************************************
ok: [localhost] => (item=None)

TASK [compatibility-layer : Publish transaction] ************************************************************************************************
skipping: [localhost]

PLAY RECAP **************************************************************************************************************************************
localhost                  : ok=15   changed=8    unreachable=0    failed=0    skipped=3    rescued=0    ignored=0
```
